{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": []
    }
   ],
   "source": [
    "#!/bin/bash \n",
    "\n",
    "### Change these settings as needed ###\n",
    "\n",
    "# Reads input options\n",
    "read_dir=\"/g/steinmetz/project/IESY/sequencing/Data/PacBio/XEMBL.20161504.PACBIO_DATA\"\n",
    "read_file=\"XEMBL_20160408_RS42150_PL100065304A-1_H01-1_lbc1.fastq \"\n",
    "read_ID=\"JS94\"\n",
    "read_ID_append=\"-A1-lowCoverage_bestn_2\"\n",
    "seq_tech=\"PacBio\"\n",
    "\n",
    "# Reference input options\n",
    "ref_dir=\"/g/steinmetz/genome/Saccharomyces_cerevisiae/\"\n",
    "ref_file=\"S288c_SynIXR/SynIXR_S299C_combined.fsa\"\n",
    "\n",
    "# Results output options\n",
    "result_dir=\"/g/steinmetz/project/IESY/sequencing/Results/PacBio/\"\n",
    "\n",
    "# Chromosome of interest\n",
    "chr_ID=\"gi|346228209|gb|JN020955.1|\" #Chr ID for collecting reads of interest\n",
    "chr_name=\"SynIXR\"\n",
    "\n",
    "# Aligner arguments\n",
    "bestn=\"2\" #number of returned hits\n",
    "nproc=\"16\" #number of spawned processes\n",
    "\n",
    "# Samtools arguments\n",
    "mem_alloc=\"20000000000\" #in bytes\n",
    "\n",
    "# Assembler arguments\n",
    "genome_size=\"155280\" #in bytes\n",
    "seq_lib_type=\"pacbio-raw\" #library input type\n",
    "use_grid=\"0\" #attach to grid\n",
    "non_default=true #true will invoke the non-default settings below (useful for low coverage assemblies)\n",
    "corMhapSensitivity=\"high\"\n",
    "corMinCoverage=\"2\"\n",
    "errorRate=\"0.035\"\n",
    "minOverlapLength=\"100\"\n",
    "corMaxEvidenceErate=\"0.3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### Run pipeline ###\n",
    "\n",
    "# Read input full path\n",
    "read_in=\"$read_dir/Demux/$read_ID/$read_file\"\n",
    "\n",
    "# Reference input full path\n",
    "ref_in=$ref_dir$ref_file\n",
    "\n",
    "# Sample ID\n",
    "sample_ID=$read_ID$read_ID_append\n",
    "\n",
    "# Aligned reads output\n",
    "align_append=\"_${seq_tech}_all_reads_aligned\"\n",
    "align_out=$sample_ID$align_append\n",
    "\n",
    "# Extracted reads output\n",
    "extract_append=\"_${seq_tech}_${chr_name}_only_sorted\"\n",
    "\n",
    "# Create results output directories\n",
    "cd $result_dir\n",
    "mkdir ./$sample_ID\n",
    "mkdir ./$sample_ID/Ref_align\n",
    "mkdir ./$sample_ID/Ref_align/All_reads\n",
    "mkdir ./$sample_ID/Canu_assembly\n",
    "mkdir ./$sample_ID/Canu_assembly/${chr_name}_only\n",
    "\n",
    "# Align reads using BLASR\n",
    "cd $result_dir$sample_ID/Ref_align/All_reads\n",
    "~/.linuxbrew/bin/blasr $read_in $ref_in -bestn $bestn -sam -nproc $nproc -out $align_out.sam & wait\n",
    "\n",
    "echo \"Completed collecting reads\"\n",
    "\n",
    "# Convert to bam, sort and index\n",
    "cd $result_dir$sample_ID/Ref_align/All_reads\n",
    "/g/software/bin/samtools view -b -S $align_out.sam > $align_out.bam\n",
    "/g/software/bin/samtools sort -m $mem_alloc $align_out.bam -o ${align_out}_sorted.bam\n",
    "/g/software/bin/samtools index ${align_out}_sorted.bam\n",
    "\n",
    "# Extract reads mapping to chromosome of interest\n",
    "/g/software/bin/samtools view -b ${align_out}_sorted.bam \"$chr_ID\" > $sample_ID$extract_append.bam\n",
    "\n",
    "# Convert result to fasta\n",
    "/g/software/bin/samtools view $sample_ID$extract_append.bam | awk '{OFS=\"\\t\"; print \">\"$1\"\\n\"$10}' - > \\\n",
    "$sample_ID$extract_append.fasta\n",
    "\n",
    "# Set JAVA environment for the assembler\n",
    "cd /g/steinmetz/project/IESY/Tools/environ\n",
    ". set_java-1.8 \n",
    "\n",
    "# Switch between default and non-default settings for the assembly\n",
    "if [ $non_default == true ]; \n",
    "    then\n",
    "        low_coverage_settings=\"corMhapSensitivity=$corMhapSensitivity corMinCoverage=$corMinCoverage \\\n",
    "        errorRate=$errorRate minOverlapLength=$minOverlapLength corMaxEvidenceErate=$corMaxEvidenceErate\"\n",
    "        echo \"Using non-default settings for the assembly\"\n",
    "    else\n",
    "        low_coverage_settings=\"\"\n",
    "        echo \"Using default settings for the assembly\"\n",
    "fi\n",
    "\n",
    "# Change to output directory for assembly\n",
    "cd /g/steinmetz/project/IESY/Tools/canu/Linux-amd64/bin/\n",
    "\n",
    "# Run CANU assembler\n",
    "./canu -p ${sample_ID}_${chr_name}_only -d \"$result_dir$sample_ID/Canu_assembly/${chr_name}_only\" \\\n",
    "genomeSize=$genome_size -$seq_lib_type \"$result_dir$sample_ID/Ref_align/All_reads/$sample_ID$extract_append.fasta\" \\\n",
    "useGrid=$use_grid $low_coverage_settings\n",
    "\n",
    "echo \"Completed assembly\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
